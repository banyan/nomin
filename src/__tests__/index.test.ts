import path from 'path';
import execa from 'execa';
import fse, { pathExists } from 'fs-extra';
import { format } from 'date-fns';
import tempy from 'tempy';
import assert from 'assert';

describe('nomin', () => {
  const tmpDir = tempy.directory();
  const bin = path.join(__dirname, '../../bin/run');
  const fixtureDir = path.join(__dirname, 'fixture');

  describe('nomin new', () => {
    it('creates a new file', async () => {
      process.chdir(tmpDir);
      await fse.copy(fixtureDir, tmpDir);
      const title = 'foo';
      const { stdout, stderr } = await execa(bin, ['new', title]);
      expect(stdout).toEqual('');
      expect(stderr).toEqual('');
      const expectedFile = `${tmpDir}/posts/${format(
        new Date(),
        'yyyy-MM-dd',
      )}-${title}.md`;
      expect(await pathExists(expectedFile)).toEqual(true);
    });

    it('shows a message if title is missing and returns exit code as 1', async () => {
      process.chdir(tmpDir);
      await fse.copy(fixtureDir, tmpDir);
      try {
        await expect(await execa(bin, ['new'])).rejects.toThrow();
      } catch (err) {
        // eslint-disable-next-line jest/no-try-expect
        expect(err.stdout).toMatchInlineSnapshot(`"nomin new requires title"`);
      }
    });
  });

  describe('nomin', () => {
    beforeAll(async () => {
      process.chdir(tmpDir);
      await fse.copy(fixtureDir, tmpDir);
      const { stdout, stderr } = await execa(bin);
      assert(stdout === '');
      assert(stderr === '');
    });

    describe('index.html', () => {
      it('creates index.html', async () => {
        const expectedFile = `${tmpDir}/public/index.html`;
        expect(await pathExists(expectedFile)).toEqual(true);
        const post = fse.readFileSync(expectedFile, 'utf8');
        expect(post).toMatchSnapshot();
      });
    });

    describe('post', () => {
      it('creates post html', async () => {
        const expectedFile = `${tmpDir}/public/2020/04/25/1/index.html`;
        expect(await pathExists(expectedFile)).toEqual(true);
        const post = fse.readFileSync(expectedFile, 'utf8');
        expect(post).toMatchSnapshot();
      });
    });

    describe('atom.xml', () => {
      it('creates atom.xml', async () => {
        const expectedFile = `${tmpDir}/public/atom.xml`;
        expect(await pathExists(expectedFile)).toEqual(true);
        const { stdout } = await execa.command(
          `grep -v updated ${expectedFile}`,
        );
        expect(stdout).toMatchSnapshot();
      });
    });

    describe('archive/index.html', () => {
      it('creates archive/index.html', async () => {
        const expectedFile = `${tmpDir}/public/archive/index.html`;
        expect(await pathExists(expectedFile)).toEqual(true);
        const archive = fse.readFileSync(expectedFile, 'utf8');
        expect(archive).toMatchSnapshot();
      });
    });

    describe('copy static', () => {
      it('copies from static to public', async () => {
        const expectedFile = `${tmpDir}/public/foo.txt`;
        expect(await pathExists(expectedFile)).toEqual(true);
      });
    });
  });
});
