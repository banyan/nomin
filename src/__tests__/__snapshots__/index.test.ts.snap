// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nomin nomin archive/index.html creates archive/index.html 1`] = `
"<html>
  <head>
    <title> - blog title</title>
    <link
      href=\\"/atom.xml\\"
      rel=\\"alternate\\"
      title=\\"blog title\\"
      type=\\"application/atom+xml\\"
    />
  </head>

  <body>
    <div>
      <header><a href=\\"/archive\\">archive</a></header>
      <h1>blog title</h1>

      <div>
        
        <ul>
          <li>
            <span class=\\"time\\">
              <time>Apr 26th, 2020</time>
            </span>
            <span>
              <a href=\\"/posts/second-post-without-permalink\\">second post without permalink</a>
            </span>
          </li>
        </ul>
        
        <ul>
          <li>
            <span class=\\"time\\">
              <time>Apr 25th, 2020</time>
            </span>
            <span>
              <a href=\\"/2020/04/25/1\\">first post with permalink</a>
            </span>
          </li>
        </ul>
        
      </div>
    </div>
  </body>
</html>
"
`;

exports[`nomin nomin atom.xml creates atom.xml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<feed xmlns=\\"http://www.w3.org/2005/Atom\\">

  <title><![CDATA[blog title]]></title>
  <link href=\\"https://example.com/atom.xml\\" rel=\\"self\\"/>
  <link href=\\"https://example.com/\\"/>
  <id>http://example.com</id>
  <author>
    <name><![CDATA[blog title]]></name>
  </author>
  <generator uri=\\"https://github.com/banyan/nomin/\\">nomin</generator>


  <entry>
    <title type=\\"html\\"><![CDATA[second post without permalink]]></title>
    <link href=\\"https://example.com/posts/second-post-without-permalink\\" />
    <id>http://example.com/posts/second-post-without-permalink</id>
    <content type=\\"html\\"><![CDATA[
<p>hi, this is a second post without permalink.</p>

]]></content>
  </entry>

  <entry>
    <title type=\\"html\\"><![CDATA[first post with permalink]]></title>
    <link href=\\"https://example.com/2020/04/25/1\\" />
    <id>http://example.com/2020/04/25/1</id>
    <content type=\\"html\\"><![CDATA[
<p>hi, this is a first post with permalink.</p>

]]></content>
  </entry>


</feed>"
`;

exports[`nomin nomin index.html creates index.html 1`] = `
"<html>
  <head>
    <title>blog title</title>
    <link
      href=\\"/atom.xml\\"
      rel=\\"alternate\\"
      title=\\"blog title\\"
      type=\\"application/atom+xml\\"
    />
  </head>

  <body>
    <div>
      <header><a href=\\"/archive\\">archive</a></header>
      <h1>blog title</h1>
      <div>
        <div>
          <h1 class=\\"entry-title\\">second post without permalink</h1>
        </div>

        <div>
          <span class=\\"time\\">
            <time>Apr 26th, 2020</time>
          </span>
          <div class=\\"main-content\\">
            <p>hi, this is a second post without permalink.</p>

          </div>
        </div>

        <div class=\\"pagination\\">
           
        </div>
      </div>
    </div>
  </body>
</html>
"
`;

exports[`nomin nomin post creates html with permalink if permalink is given 1`] = `
"<html>
  <head>
    <title>first post with permalink - blog title</title>
    <link
      href=\\"/atom.xml\\"
      rel=\\"alternate\\"
      title=\\"blog title\\"
      type=\\"application/atom+xml\\"
    />
  </head>

  <body>
    <div>
      <header><a href=\\"/archive\\">archive</a></header>
      <h1>blog title</h1>
      <div>
        <div>
          <h1 class=\\"entry-title\\">first post with permalink</h1>
        </div>

        <div>
          <span class=\\"time\\">
            <time>Apr 25th, 2020</time>
          </span>
          <div class=\\"main-content\\">
            <p>hi, this is a first post with permalink.</p>

          </div>
        </div>

        <div class=\\"pagination\\">
           
          <a href=\\"/posts/second-post-without-permalink\\">later »</a>
          
        </div>
      </div>
    </div>
  </body>
</html>
"
`;

exports[`nomin nomin post creates html with title if permalink is not given 1`] = `
"<html>
  <head>
    <title>second post without permalink - blog title</title>
    <link
      href=\\"/atom.xml\\"
      rel=\\"alternate\\"
      title=\\"blog title\\"
      type=\\"application/atom+xml\\"
    />
  </head>

  <body>
    <div>
      <header><a href=\\"/archive\\">archive</a></header>
      <h1>blog title</h1>
      <div>
        <div>
          <h1 class=\\"entry-title\\">second post without permalink</h1>
        </div>

        <div>
          <span class=\\"time\\">
            <time>Apr 26th, 2020</time>
          </span>
          <div class=\\"main-content\\">
            <p>hi, this is a second post without permalink.</p>

          </div>
        </div>

        <div class=\\"pagination\\">
          
          <a href=\\"/2020/04/25/1\\" rel=\\"next\\">« earlier</a>
           
        </div>
      </div>
    </div>
  </body>
</html>
"
`;
